src/bio.lisp

```diff
@@ -104,14 +173,52 @@
         do
     (setf (cffi:mem-ref buf :unsigned-char i) (read-byte *socket*))
     (incf i))
-    #+(or)
+    i
     (when (zerop i) (set-retry-read bio)))
-      (end-of-file ()))
+      (end-of-file ()
+        ;; could this be part of serious condition?
+        (clear-retry-flags bio)
+        (setf (cffi:mem-ref buf :unsigned-char i) 0))
+      (serious-condition ()
+        (clear-retry-flags bio)
+        ;; we could set BIO_set_retry_reason() if we defined
+        ;; codes. Could be useful for timeouts, but not implemented now.
+        ))
+    i))
+
+(cffi:defcallback lisp-gets :int ((bio :pointer) (buf :pointer) (n :int))
+  (let ((i 0))
+    (handler-case
+        (unless (or (cffi:null-pointer-p buf) (null n))
+          (clear-retry-flags bio)
+          (when (or *blockp* (listen *socket*))
+            (setf (cffi:mem-ref buf :unsigned-char i) (read-byte *socket*))
+            (incf i))
+          (loop
+            with char
+            and exit = nil
+            while (and (< i n)
+                       (null exit)
+                       (or (null *partial-read-p*) (listen *socket*)))
+            do
+               (setf char (read-byte *socket*)
+                     exit (= char 10))
+               (setf (cffi:mem-ref buf :unsigned-char i) char)
+               (incf i)))
+      (serious-condition ()
+        (clear-retry-flags bio)))
+    (unless (>= i n)
+      (setf (cffi:mem-ref buf :unsigned-char i) 0))
```

Comment:
We should always 0-terminate. This also means loop while i <  n - 1.

Reply:
Agreed.